XP-422 ai metrics

janpetschexain opened the pull request:
### References

- [XP-422](https://xainag.atlassian.net/browse/XP-422)

### Summary

- update the participant state machine to use the new ai metrics object
- update the keras and pytorch example and documentation

### Are there any open tasks/blockers for the ticket?

- note that these changes are interdependent with the merge requests of [xain-fl](https://github.com/xainag/xain-fl/pull/216) and [xain-proto](https://github.com/xainag/xain-proto/pull/3) and need to be merged simultaneously to avoid breaking stuff

---

### Reviewer checklist

*Reviewer agreement:*

* Reviewers assign themselves at the start of the review.
* Reviewers do **not** commit or merge the merge request.
* Reviewers have to check and mark items in the checklist.

**Merge request checklist**

- [x] Conforms to the merge request title naming `XP-XXX <a description in imperative form>`.
- [x] Each commit conforms to the naming convention `XP-XXX <a description in imperative form>`.
- [x] Linked the ticket in the merge request title or the references section.
- [x] Added an informative merge request summary.

**Code checklist**

- [x] Conforms to the branch naming `XP-XXX-<a_small_stub>`.
- [ ] Passed scope checks.
- [ ] Added or updated tests if needed.
- [x] Added or updated code documentation if needed.
- [ ] Conforms to Google docstring style.
- [ ] Conforms to XAIN structlog style.

wilk10 commented on docs/examples/pytorch.md:
```diff
@@ -109,162 +123,75 @@ def train_n_epochs(self, trainloader, number_of_epochs: int) -> None:
 
-In the tutorial we  use the standard data  transformation and loading  as it is described in the PyTorch documentation. 
-The ony difference is that we initiate each Participant with  a randomised training set. 
+In the tutorial we use the standard data transformation and loading as it is described in the PyTorch documentation. The ony difference is that we initiate each Participant with a randomised training set.
```
**only** difference

janpetschexain replied:
fixed.

wilk10 commented on xain_sdk/participant_state_machine.py:
```diff
@@ -141,15 +149,23 @@ def training_round(channel: Channel, participant: Participant) -> None:
-        channel=channel, weights=weights, number_samples=number_samples, metrics=metrics
-    )
+    # data validation
+    if not isinstance(weights, List) or not all(isinstance(weight, ndarray) for weight in weights):
```
maybe we could change this to be more readable?

`all_ndarray = all(isinstance(weight, ndarray) for weight in weights)`
`if not isinstance(weights, List) or not all_ndarray:`
`    raise ...`

janpetschexain replied:
the problem with the suggestion is, that python couldn't benefit from boolean short-circuiting. in case of `if a and b: ...` it would never evaluate `b` if `a` is already false. but in case of
```
a = ...
b = ...
if a and b: ...
```
it would always evaluate `a` and `b` first before attempting to evaluate the if-clause.

i agree that longer if-clauses are less readable, but we unfortunately can't have boolean short-circuits and readability at the same time. therefor i recommend to not split it into multiple lines.

janpetschexain replied:
> thanks! looks good but just to double-check can you please point to the functions that you changed within the scope (excluding pylint stuff, i mean)? the `Participant` base class and then `start_training` and `end_training` in `participant_state_machine`?

sure, the `train_round` of the participant base class, the `end_training` of the participant state machine, the `train_round` of the tensorflow and pytorch examples as well as their markdown documentation files were changed to use the new metrics object.

janpetschexain replied:
changed.

wilk10 commented on xain_sdk/participant_state_machine.py:
```diff
@@ -141,15 +149,23 @@ def training_round(channel: Channel, participant: Participant) -> None:
+        not isinstance(metrics, Dict)
+        or not all(isinstance(key, str) for key in metrics.keys())
+        or not all(isinstance(value, ndarray) for value in metrics.values())
+    ):
```
same here, maybe if broken up into three variables and if checks all of them together, the naming of the variable will help readability

janpetschexain replied:
changed.

PanicButtonPressed commented:
Can you also modify the [README.md](https://xainag.atlassian.net/browse/XP-432) of the repository to reflect the latest changes?

janpetschexain commented:
> looks good! just one more change please, there are two `is_not_ndarray` variables now, maybe we can give two different names and add the typing to the second one? thanks!

changed names, added type.
